### ISSUE TITLE
TypeError: Invalid attempt to destructure non-iterable instance

### ISSUE THREAD

**Issue by astronaut-wannabe** [handsontable/handsontable#4637]
### Description
The [`afterChange`][0] hook is supposed to receive a 2D array of changes.
Somewhere a callback is passing in a 1D array resulting in:
```
TypeError: Invalid attempt to destructure non-iterable instance
```

I think this occurs when both the sorting plugin and the filter plugin are on and is related to the ObserveChanges plugin (since this only happens when a change is triggered via the source data)
 
### Steps to reproduce
1. open the following code in a browser 
```html
</style><!-- Ugly Hack due to jsFiddle issue -->

<script src="https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js"></script>
<link type="text/css" rel="stylesheet" href="https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.css">

<div id="container"></div>
<div id="example1" class="hot handsontable htRowHeaders htColumnHeaders"></div>
<script>
var example = document.getElementById('example1');
var hot;

hot = new Handsontable(container, {
  data: Handsontable.helper.createSpreadsheetData(50, 15),
  width: 500,
  height: 300,
  rowHeaders: true,
  colHeaders: true,
  columnSorting: true,
  sortIndicator: true,
  dropdownMenu: true,
  filters: true
});
</script>
```
2. Open browser developer tools and enable "pause on uncaught exceptions"
<img width="371" alt="screen shot 2017-11-29 at 3 38 45 pm" src="https://user-images.githubusercontent.com/13190980/33392659-76b786c6-d51b-11e7-95ab-067b0a0d0727.png">

3. in the console, run this code:
```javascript
var row_data = hot.getSourceDataAtRow(0);
row_data[0] = "an exception will raise";
````
4. wait a bit for the dirty check to fire the hook
5. the exception will raise:
```javascript
TypeError: Invalid attempt to destructure non-iterable instance handsontable.full.min.js:405:297347
	s</< https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:405:297347
	value/< https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:405:304142
	u https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:26:1381
	value https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:405:304107
	value/< https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:405:301510
	value https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:26:28887
	i/this.runHooks https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:400:88641
	replace https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:405:231047
	value/< https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:405:231147
	u https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:26:1381
	value https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:405:231109
	value/< https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:405:229914
	runLocalHooks/< https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:26:96664
	u https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:26:1381
	runLocalHooks https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:26:96610
	value https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:405:233200
	value/this.observer< https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:405:232999
	h https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:400:213572
	d https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:400:212353
	p/n.next< https://docs.handsontable.com/pro/1.14.3/bower_components/handsontable-pro/dist/handsontable.full.min.js:400:212422
```
### Demo
https://jsfiddle.net/vhrxp5ob/
![demo](https://user-images.githubusercontent.com/13190980/33393747-f394a25c-d51e-11e7-9c60-1cf11d3a9522.gif)

### Your environment
* Handsontable version: Pro 1.14.3
* Browser Name and version: 
  * Chrome Version 62.0.3202.94 (Official Build) (64-bit)
  * Firefox 58.0b7 (64-bit)
* Operating System: macOS Sierra 10.12.6 (16G1036)

[0]:https://docs.handsontable.com/pro/1.14.3/Hooks.html#event:afterChange

**Comment by swistach** [comment]
Hi @astronaut-wannabe,
thank you! You can disable `observeChanges` in your settings object. Even without `observeChanges`, `columnSorting` should work properly.

@handsontable/core we have to make a decision, what we want to do with a `observeChanges` plugin. Maybe this feature is unnecessary and we can remove it from `Handsontable`?

I mark this issue as a bug to discuss.

**Comment by astronaut-wannabe** [comment]
@swistach no problem. I applied the change in #4638 to my local copy of HOT to fix the error, but it is good to know that disabling `observeChanges` is an option

**Comment by swistach** [comment]
Released in v4.0.0
